traits/src/location.rs

```diff
@@ -22,6 +22,9 @@ impl Parse for MultiLocation {
 			(1, _) => Some(MultiLocation::parent()),
 			// children parachain
 			(0, Some(Parachain(id))) => Some(MultiLocation::new(0, X1(Parachain(*id)))),
+			// self reserve
+			// All asset that is not a children parachain is a self reserved asset
```

Comment:
One of `chain_part` use case is to check user errors on `dest` param in xtokens calls.

https://github.com/open-web3-stack/open-runtime-module-library/blob/4a66b299037cc3997689538f82847785f9afa65d/xtokens/src/lib.rs#L729

This will make any random `dest` has valid chain part.

It will also make any multi assets has valid reserve. https://github.com/open-web3-stack/open-runtime-module-library/blob/4a66b299037cc3997689538f82847785f9afa65d/xtokens/src/lib.rs#L753

Reply:
The core issue here is that the new reanchoring logic encourage a "relative path" point of view of `MultiLocation`. For example, if the current working parachain ID is `1`. Then, the following two MultiLocations should be equivalent:
```rust
../Parachain(1)/xxx
```
```rust
xxx
```
The previous ORML design and tests are leaning towards always using the first case. This is fine, but I think using the second case should be encouraged since it more simple.

> It will also make any multi assets has valid reserve.
Not really, see the updated new test. 

I think at end of the day, it really matters what is the opinionated view of how a local asset's location should be represented. In my view (seems also by Parity), your own parachain balance should be:
```rust
here()
```
instead of 
```rust
../Parachain(your_own_id)
```
The asset on your own should be
```rust
./asset_location
```
instead of:
```rust
../Parachain(your_own_id)/asset_location
```

Reply:
ORML uses "absolute" location path on purpose. One of the benefit is to be compatible with token location register module we planned. It's a design choice, instead of which point of view we preferred to see assets on our own parachain.

> I think using the second case should be encouraged since it simpler.

I don't see how it makes implementation and integration simpler, but it needs to cover more edge cases.

> I think at end of the day, it really matters what is the opinionated view of how a local asset's location should be represented. In my view (seems also by Parity), your own parachain balance should be:

XCM executor is designed to be flexible and in the end it's parachains choice to decide how to convert between multi location and their local currency. Not sure about what you mean by Parity's view.

If ORML impl changed to relative location path design, the way how params are used in xtokens would be very different. It's not just how local asset are represent but also foreign locations in `dest` and `assets`. It would be a breaking change.

Reply:
One way that we could fullfil both requests is to introduce a new `pub trait ReserveCalculator` like 

```
pub trait ReserveCalculator {
	/// Calculates a reserve location of a multiasset.
	fn calculate_reserve(&MultiAsset) -> Option<MultiLocation>;
}
```

And add a new associated type in the config of orml_xtokens
```
/// Reserve Calculator
type ReserveCalculator: ReserveCalculator;
```

That way xtokens is not tied to a specific way of seeing self tokens (absolute vs relative) and each runtime can configure it indepentently

Reply:
we got `Reserve` trait already, if we want expose a trait, I think we could use `Reserve` and make some adjustment.

https://github.com/open-web3-stack/open-runtime-module-library/blob/4a66b299037cc3997689538f82847785f9afa65d/traits/src/location.rs#L43-L46

Reply:
That works too